Bug 1210305 lock screen

harman-ankitgupta opened the pull request:
@Ashish-Developer @pgmanutd

TatianaMigacheva commented:
For next commit please update 'Commit Message' according to template (http://stteamsite.symphonyteleca.com/redsquare/Wiki/Commit_message_template.aspx).

harman-jenkins commented:
Results: \runn14lsp002\shared\builds\pull-requests\gaia_pr22\

sergkashin commented on apps/system/js/notifications.js:
```diff
@@ -403,44 +402,57 @@ var NotificationScreen = {
-      var icon = document.createElement('img');
-      icon.src = detail.icon;
+    if (detail.appIcon) {
+    var icon = document.createElement('img');
```
wrong indentation

sergkashin commented on apps/system/js/notifications.js:
```diff
@@ -456,10 +468,31 @@ var NotificationScreen = {
     message.appendChild(messageContent);
     notificationNode.appendChild(message);
+    } else {
+    // in case if notification is already present on the screen, its counter
```
wrong indentation in all "else {}" statement

egorlevichev commented on apps/system/js/notifications.js:
```diff
@@ -403,44 +402,57 @@ var NotificationScreen = {
+    // 'detail' object, name of the sender is coming. in other cases title text
+    // in 'detail' object is appropriate.
+    var cIndex = this.unReadNotificationIndex(manifestURL);
+    if (cIndex === -1) {
```
"It looks like the closing brace for this if clause is 60 lines later! And you haven't indented the intervening 60 lines correctly, so the code becomes very hard to read." (C) David from PR https://github.com/mozilla-b2g/gaia/pull/32749.
Please, in such cases ((un)wrapping of existent code with `if` or `switch` or whatever) don't forget to properly indent such blocks and if nothing except indentation is changed in such cases, you can just put comment in PR mentioning this, to speed up review.

harman-ankitgupta replied:
I have done changes in that 60 lines of code as well, if i indent complete block of lines then it will be very hard to review the changes for David. Once David review this code and approve it then i will indent this block in any future lock screen commits. 

This is inline as per the second part from David's comments last time => "(I know I complained on an earlier version about unnecessary changes making the diff too big. But I think those were about accidental, incorrect indentation changes. In this case, the code must be indented in order to be legible, even if it adds 60 lines to your diff. (When something like that happens, you can always add a comment here on github letting me know that nothing has changed other than indentation.)"

egorlevichev replied:
This still wrong, so even if I'll approve it, this version of code will not be approved by David since it violates Mozilla coding style guide, so it won't be landed and thus there will be no 'future lock screen commits' until this one will be landed.

Here is some kind of explanation about indentations and the way of making PRs:
1. Initial comments from David were about changing **only** indentation of not changed code, especially (as it was in original commit) wrong changes which lead to shifting complete functions to the left. This is wrong.
2. Next comment about this particular line was in couple of words: if you adding new nesting level (wrap some code with `'{}'`) - then you **shall update indentation**.
   - If nothing except indentation is changed after wrapping the code - you can leave comment in PR about this change - it will help reviewers understand your code.
   - If you changing some code in wrapped block, then it's OK to not leave comment or to leave comments inline the code explaining what have being changed (if you feel that such comments will help to better understand your changes) _NOTE: only way to notify reviewers about your exact changes in such case is to make 2 separate commits: first one - changing indentation, second - making additional changes in indented code. After review such commits shall be squashed into single one going to integration._ 
3. According to your commit message this PR contains 2 parts: 1 - fixes of review comments from mozilla-b2g#32749 and 2 - fixes/implementation of several new functionalities. For better understanding and for speeding up reviews, could you please in the future don't mix everything into one PR. PR1 - for comments & PR2 - for everything else could be perfect.

egorlevichev commented on apps/system/js/notifications.js:
```diff
@@ -84,7 +84,6 @@ var NotificationScreen = {
     // screen
     this.externalNotificationsCount = 0;
     this.unreadNotifications = [];
```
I think that `Map` could fit better for storing key-value pairs

egorlevichev replied:
One more point regarding re-use of this variable: in original version it's used to show/hide ambient indicator (tiny blue line under the status bar) so this array will be forcibly cleaned in `function ns_hideNotificationIndicator()`. Also this array is expected to store only 'unread' notifications, i.e. ones which were not seen by user, so it's expected that after, for example, opening list of notifications all of them became read, but continue exist until the message/call/whatever won't be actually opened/read in proper application. I can see here 2 possible solutions:
1. In case if ambient indicator is not expected to be in RedSquare (here we need confirmation for UX team) - completely remove related functionality and re-use `unreadNotifications`.
2. Otherwise just create new variable for storing notifications (e.g. #21, where similar counters were implemented to be able to show amounts of notifications for individual applications)

egorlevichev commented on apps/system/js/notifications.js:
```diff
@@ -403,44 +402,57 @@ var NotificationScreen = {
     notificationNode.dataset.obsoleteAPI = 'false';
     if (typeof detail.id === 'string' &&
-        detail.id.indexOf('app-notif-') === 0) {
+      detail.id.indexOf('app-notif-') === 0) {
```
wrong indentation here, originally it was properly indented (one part of `if` condition under another) just like functions with multiple arguments.

egorlevichev commented:
By the way, since this PR doesn't have any description except title, it's not clear for me if this PR is a fix of comments from https://github.com/mozilla-b2g/gaia/pull/32749 or it's some new functionality implemented on top of 32749 with (or without) taking into account the comments?

harman-ankitgupta commented:
@egorlevichev I have already added a comment that issues related to David's comments are fixed.

egorlevichev commented on apps/system/js/notifications.js:
```diff
@@ -624,18 +657,18 @@ var NotificationScreen = {
   },
 
-  addUnreadNotification: function ns_addUnreadNotification(id, skipUpdate) {
+  addUnreadNotification: function ns_addUnreadNotification(appManifestURL, skipUpdate) {
```
actually you don't need to change signature of this function, you are using value of `appManifestURL` as an identifier for notifications counter.

harman-ankitgupta replied:
I have changed the signature of the function to make it self-expalantory that instead of id ,appManifestURL is being used now.

egorlevichev commented on apps/system/js/notifications.js:
```diff
@@ -624,18 +657,18 @@ var NotificationScreen = {
       return;
     }
-    this.unreadNotifications.push(id);
+    this.unreadNotifications.push({name: appManifestURL, count: 1});
```
better to use 'appId' or something similar for the first attribute name instead of 'name' used now.

harman-ankitgupta replied:
will replace this with Map. So this code will be removed.

egorlevichev commented on apps/system/js/notifications.js:
```diff
@@ -768,9 +801,19 @@ var NotificationScreen = {
+  /*
+   * Returns the index of the object for the given manifestURL
+   */
+  unReadNotificationIndex: function ns_unReadNotificationIndex(manifestUrl) {
```
unReadNotificationIndex -> unreadNotificationIndex

harman-ankitgupta replied:
will remove this code as Maps are used

egorlevichev commented on apps/system/js/notifications.js:
```diff
@@ -456,10 +468,31 @@ var NotificationScreen = {
+      var notificationNode = document.querySelector('.notification');
+      var countVal = this.unreadNotifications[cIndex].count;
+      this.unreadNotifications[cIndex].count = ++countVal;
+      document.getElementById(manifestURL).querySelector('.detail-count').textContent = countVal;
```
you are calling the same function `document.getElementById(manifestURL)` 3 times in this block, it will be more efficient if you call it once, assign returned value to the variable and then use that variable in all 3 places.

harman-ankitgupta replied:
will refactor it

egorlevichev commented on apps/system/js/notifications.js:
```diff
@@ -403,44 +402,57 @@ var NotificationScreen = {
     notificationNode.setAttribute('role', 'link');
-
+    notificationNode.setAttribute('id', manifestURL);
+    notificationNode.setAttribute('tabindex', '0');
```
if I understand correctly each newly created `notificationNode` will have tabIndex==0, is it expected to increment these indexes for previously added nodes?

harman-ankitgupta replied:
To make the element focusable via keyboard ''tabIndex" should be set to zero.

egorlevichev replied:
ok

egorlevichev commented on apps/system/js/notifications.js:
```diff
@@ -403,44 +402,57 @@ var NotificationScreen = {
-    title.textContent = detail.title;
+    // in case of text message, title in 'detail' object is name of the sender
+    // but as per the UX title will be 'new messages'.
+    if(detail.manifestURL === smsManifestUrl) {
```
you are making separate processing only for sms messages, but as per UI spec there at least 3 different types of 'titles' for notifications. Maybe it will be better to move customization of `notificationNode` according to type of application to the separate utility function?

harman-ankitgupta replied:
As per UX there are 3 different kind of notifications explained:
1- Message (Which i have handled above)
2- Missed Call (Default implementation handled it properly.)
3- Third party (UX description for title is not explanatory enough for the user, so I kept is as default)

We might need to revisit this part later.

egorlevichev commented on apps/system/js/notifications.js:
```diff
@@ -456,10 +468,31 @@ var NotificationScreen = {
+      document.getElementById(manifestURL).querySelector('.detail-count').textContent = countVal;
+      var message = document.getElementById(manifestURL).querySelector('.detail-content').textContent;
+      if(detail.manifestURL === smsManifestUrl) {
+        message = message + "; " + detail.title + ': ' + detail.text;
```
I'm not sure about requirements for lockScreen notifications, but for homescreen there is a requirement to decrement counter if user reads for example one of sms messages, so in case of concatenated text from several messages you will need to extract parts of string which belongs to already read message from this common string. This is one more reason to store required notifications data not only in html nodes, but, probably in additional array of objects.

egorlevichev commented on apps/system/lockscreen/js/lockscreen_inputpad.js:
```diff
@@ -107,27 +98,14 @@
-        // handlePassCodeInput triggers updatePassCode
-        this.handlePassCodeInput(key);
+      case 'keyup':
+        this.handleKeyEvent(evt);
```
Why not to register `handleKeyEvent(evt)` as an event listener directly on line 44 without putting it's invocation here?

harman-ankitgupta replied:
They have implemented in this way only so i did it in the same way. But I will refactor as suggested.

egorlevichev commented on apps/system/lockscreen/js/lockscreen_inputpad.js:
```diff
@@ -163,16 +137,59 @@
       var span = this.passcodeCode.childNodes[i];
       if (span) {
         if (this.states.passCodeEntered.length > i) {
+        var value = this.states.passCodeEntered.substr(i, this.states.passCodeEntered.length);
```
wrong indentation

harman-ankitgupta replied:
will refactor it

egorlevichev commented on apps/system/lockscreen/js/lockscreen_notification_builder.js:
```diff
@@ -59,97 +59,19 @@
 
     node.addEventListener('click',
-      this.requestHighlight.bind(this, node));
+      this.requestActivate.bind(this, node));
```
this code seems never will be called, since there won't be any `click` events in non-touch environment.

harman-ankitgupta replied:
I have tested it in emulator, click event is getting fired when "Ok" button is pressed.

egorlevichev commented on apps/system/lockscreen/js/lockscreen_state_manager.js:
```diff
@@ -411,14 +413,46 @@
+        appID = detail.id;
         break;
       case 'lockscreen-request-unlock':
+      // if app is launched from the lockscreen notification screen then it will
```
I agree with David at this point - this code shouldn't even be here.
launch of an application currently performed in 'visibility_manager.js' as a response to data provided with `lockscreen-request-unlock` event.

harman-ankitgupta replied:
Will remove this code

egorlevichev commented on apps/system/lockscreen/js/lockscreen_state_manager.js:
```diff
@@ -453,6 +487,11 @@
         this.onSecureAppClosing();
         break;
+      case 'keyup':
+        if(this.lockScreenStates.passcodeEnabled && (evt.keyCode >= KEY_CODE_ZERO && evt.keyCode <= KEY_CODE_NINE)) {
```
What will happen if passcode lock disabled? How user expected to dismiss screen lock in such case?

harman-ankitgupta replied:
This requirement is under discussion with UX team.

egorlevichev replied:
As soon as decision most probably won't be made regarding this case, you shall leave ability to dismiss lockscreen (when 'pin lock' disabled) for example by pressing on any hardware key. Otherwise your feature will make device/emulator really hard to use (at least for mentioned configuration: {lock: ON; pin: OFF}).

egorlevichev commented on apps/system/js/notifications.js:
```diff
@@ -403,44 +402,57 @@ var NotificationScreen = {
+    if (detail.appIcon) {
       var icon = document.createElement('img');
-      icon.src = detail.icon;
+      icon.src = detail.appIcon;
```
for missed call scenario this change will result in wrong icon displayed, see attached screenshot: 
![screenshot from 2015-11-07 19 31 50](https://cloud.githubusercontent.com/assets/7557901/11016076/ce76c822-8586-11e5-84d9-25c3bcda3905.png)
it happens because `detail.appIcon == "app://communications.gaiamobile.org/style/icons/communications_45.png"` while `detail.icon == "app://communications.gaiamobile.org/dialer/style/icons/dialer_84.png"`

harman-ankitgupta replied:
If we install an external application to post the notification detail.icon comes empty instead of that detail.appIcon contains the path for the icon.

egorlevichev replied:
I still believe that it's not a good idea to drop **existing platform** functionality just to be able to see icon from some 3-d party app, while even the possibility of such applications installation is still under discussion for this project.
Acceptable workaround could be to check if `detail.icon` exists, and **only if it's not** try to use any other image.

harman-jenkins commented:
Harman-Jenkins: pr_review_passed

egorlevichev commented on apps/system/js/notifications.js:
```diff
@@ -403,20 +403,24 @@ var NotificationScreen = {
+    // will be the count along with the 'new message' text,but currently in
+    // 'detail' object, name of the sender is coming. in other cases title text
+    // in 'detail' object is appropriate.
+    var cIndex = this.unreadNotifications.has(manifestURL);
```
this variable shall be renamed, it's not an index anymore

egorlevichev commented on apps/system/js/notifications.js:
```diff
@@ -403,20 +403,24 @@ var NotificationScreen = {
     // timesout, so we skip updating it here, by passing a skip bool
-    this.addUnreadNotification(detail.id, true);
-
+    this.addUnreadNotification(manifestURL, true);
```
The source code with wrong indentation can't be integrated.

When I mentioned separation on 2 commits, the idea was:

1 - Initial state without any your changes

``` javascript
code_line: 1;
code_line: 2;
```

2 - Your commit number 1 (wrap this code with `{}` and add indentation)

``` javascript
if (magic) {
  code_line: 1;
  code_line: 2;
} else {
...
}
```

3 - Your commit number 2 (just to highlight changes in indented code)

``` javascript
if (magic) {
  code_line: 1;
  your_new: code-line;  // made required code changes add/modify/remove anything you need
  code_line: 2;
} else {
...
}
```

As a result of such sequence of 2 commits you will have: 1) all required changes; 2) review of them will be as simple as possible.
In your case you have second commit without must-have changes, so to be able to integrate such PR you anyway will need to create one more commit containing everything what shall be integrated.

harman-jenkins commented:
Harman-Jenkins: pr_review_passed

harman-jenkins commented:
Harman-Jenkins: pr_review_passed

egorlevichev commented on shared/js/lockscreen_slide.js:
```diff
@@ -343,12 +332,10 @@
     function lss_initializeCanvas() {
 
       this.overlay = document.getElementById(this.IDs.overlay);
-      this.area = document.getElementById(this.IDs.area);
```
after this change following attempts to use this member will fail with: `JavaScript Error: "TypeError: this.area is null"`

harman-jenkins commented:
Harman-Jenkins: pr_review_passed

harman-jenkins commented:
Harman-Jenkins: pr_review_passed

harman-jenkins commented:
Harman-Jenkins: pr_review_passed

harman-jenkins commented:
Harman-Jenkins: pr_review_passed

harman-jenkins commented:
Harman-Jenkins: pr_review_passed

harman-jenkins commented:
Harman-Jenkins: pr_review_passed

harman-jenkins commented:
Harman-Jenkins: pr_review_passed

harman-jenkins commented:
Harman-Jenkins: pr_review_passed

egorlevichev commented on apps/system/js/notifications.js:
```diff
@@ -403,64 +404,80 @@ var NotificationScreen = {
+    } else {
+      // in case if notification is already present on the screen, its counter
+      // value will be incremented.
+      var notificationNode = document.querySelector('.notification');
```
as debugging shows - it's not safe way to retrieve notification node - it will return first observed item with 'notification' in it's classList, and it easily won't be an item you created in `if()` branch above. So line 474 will cause error: `JavaScript Error: "TypeError: notificationNode.querySelector(...) is null"` and the remaining code in this function won't be processed.

harman-jenkins commented:
Harman-Jenkins: pr_review_passed

harman-jenkins commented:
Harman-Jenkins: pr_review_passed

harman-jenkins commented:
Harman-Jenkins: pr_review_passed

sergkashin commented on apps/system/js/notifications.js:
```diff
@@ -403,64 +404,80 @@ var NotificationScreen = {
+      // We need to animate the ambient indicator when the toast
+      // timesout, so we skip updating it here, by passing a skip bool
+      this.addUnreadNotification(manifestURL, true);
+      var notificationNode = document.createElement('div');
```
don't you think that using https://developer.mozilla.org/en-US/docs/Web/API/Element/insertAdjacentHTML would be much easier and prettier here instead of using of the document.createElement() function?
The HTML fragment might be defined just like it is done in app_window.js:AppWindow.prototype.view function. 
For insertAdjacentHTML() usage you can check AppWindow.prototype._render function as well as in the docs mentioned above.

sergkashin replied:
meh, never mind, looks like it was originally done like this, so ignore my comment.

harman-jenkins commented:
Harman-Jenkins: pr_review_passed

harman-jenkins commented:
Harman-Jenkins: pr_review_passed

egorlevichev commented:
:shipit:

Though it still contains number of weird cases, e.g. :
1. controls for entering PIN appears after lockscreen dismissed, and user is able to enter PIN over the homescreen.
   Steps to reproduce: 
   - unlock device by entering valid PIN;
   - start entering any digit on the homescreen)
     example screenshot: 
     ![screenshot from 2015-11-13 14 15 04](https://cloud.githubusercontent.com/assets/7557901/11144989/4ef486ee-8a12-11e5-8ed2-05e42f7b3925.png)
2. list of notifications is not restored after locking device (not sure if it's bug or feature). Steps:
   - get some notifications on the lock screen;
   - unlock device;
   - lock device;
   - there is no notifications about missed calls/messages on lock screen, though there are still unread messages and unseen missed calls

harman-ankitgupta commented:
@egorlevichev 
Case 1: this issue has been fixed and commited , please review it.
Case 2: This is not a bug, All the screen notification will be removed as soon as device is unlocked.

TatianaMigacheva commented:
Harman-Jenkins: pr_review_failed

**Lint errors:**
gjslint check:
Running gjslint...
----- FILE  :  /export/jenkins/jobs/Gaia_PR_checker/workspace/checkstyle/gaia/apps/system/js/notifications.js -----
Line 436, E:0001: Extra space before ";"
Line 453, E:0110: Line too long (81 characters).
Line 464, E:0002: Missing space after ","
Line 475, E:0110: Line too long (82 characters).
Line 477, E:0110: Line too long (82 characters).
Line 644, E:0110: Line too long (88 characters).
Line 654, E:0110: Line too long (82 characters).
Line 790, E:0110: Line too long (86 characters).
Line 794, E:0131: Single-quoted string preferred over double-quoted string.
Line 794, E:0110: Line too long (135 characters).
Line 796, E:0131: Single-quoted string preferred over double-quoted string.
Line 796, E:0110: Line too long (91 characters).
Line 805, E:0001: Extra space after "?"
Line 805, E:0110: Line too long (124 characters).
Line 807, E:0001: Extra space after "?"
Line 807, E:0110: Line too long (81 characters).
----- FILE  :  /export/jenkins/jobs/Gaia_PR_checker/workspace/checkstyle/gaia/shared/js/lockscreen_slide.js -----
Line 671, E:0236: Error parsing jsdoc type at token "(" (column: 21): Invalid identifier for (): "Function"
Line 18, E:0002: Missing space before "|"
Line 18, E:0002: Missing space before "opts"
Line 212, E:0002: Missing space before "|"
Line 212, E:0002: Missing space before "opts"
Line 243, E:0002: Missing space before "|"
Line 243, E:0002: Missing space before "options"
Line 267, E:0002: Missing space before "|"
Line 267, E:0002: Missing space before "evt"
Line 474, E:0002: Missing space before "|"
Line 474, E:0002: Missing space before "tx"
Line 602, E:0002: Missing space before "|"
Line 602, E:0002: Missing space before "pageX"
Line 603, E:0002: Missing space before "|"
Line 603, E:0002: Missing space before "pageY"
Line 636, E:0002: Missing space before "|"
Line 636, E:0002: Missing space before "tx"
Line 670, E:0002: Missing space before "|"
Line 670, E:0002: Missing space before "tx"
Line 671, E:0002: Missing space before "|"
Line 671, E:0002: Missing space before "cb"
Line 737, E:0002: Missing space before "|"
Line 737, E:0002: Missing space before "tx"
Line 888, E:0002: Missing space before "|"
Line 888, E:0002: Missing space before "tx"
Line 1059, E:0002: Missing space before "|"
Line 1059, E:0002: Missing space before "px"
Line 1094, E:0002: Missing space before "|"
Line 1094, E:0002: Missing space before "px"
Line 1120, E:0002: Missing space before "|"
Line 1120, E:0002: Missing space before "x"
Line 1121, E:0002: Missing space before "|"
Line 1121, E:0002: Missing space before "y"
Found 49 errors, including 1 new error, in 2 files (0 files OK).

Some of the errors reported by GJsLint may be auto-fixable using the script
fixjsstyle. Please double check any changes it makes and report any bugs. The
script can be run by executing:

fixjsstyle --disable 210,217,220,225 --custom_jsdoc_tags=prop,borrows,memberof,augments,exports,global,event,example,mixes,mixin,fires,inner,todo,access,namespace,listens,module,memberOf,property,requires,alias,returns -e apps/pdfjs/content,apps/pdfjs/test,apps/email/build,apps/email/built,apps/email/js/ext,apps/calendar/build,apps/calendar/js/ext,apps/camera/js/vendor,apps/camera/test,apps/privacy-panel/js/vendor,apps/system/camera,apps/keyboard/js/imes/jszhuyin/lib,apps/keyboard/js/imes/handwriting/hwr,apps/wappush/js/ext,apps/keyboard/js/layouts,apps/music/bower_components,apps/camera/bower_components -x build/r.js,build/csslinter.js,build/jsmin.js,apps/communications/contacts/oauth2/js/parameters.js,apps/calendar/js/presets.js,apps/email/js/alameda.js,apps/email/js/tmpl_builder.js,apps/clock/js/alameda.js,apps/clock/js/text_builder.js,shared/js/utilities.js,shared/js/uuid.js,apps/keyboard/js/imes/jspinyin/libpinyin.js,shared/js/setImmediate.js,shared/test/unit/test-agent-server.js,shared/elements/gaia-header/dist/gaia-header.js,shared/elements/gaia-component/gaia-component.js,shared/elements/gaia-icons/gaia-icons.js,shared/pages/import/js/parameters.js,apps/sharedtest/test/unit/setImmediate_test.js,apps/keyboard/js/imes/jsavrophonetic/jsavrophonetic.js,apps/camera/js/vendor/alameda.js,apps/camera/js/config.js,apps/gallery/js/config.js,apps/gallery/js/frame_scripts.js,apps/gallery/js/metadata_scripts.js,apps/music/js/metadata_scripts.js,apps/keyboard/js/imes/jszhuyin/tools/cook.js,apps/keyboard/build/keyboard-config.js,apps/settings/js/vendor/alameda.js,apps/findmydevice/js/lib/*,apps/findmydevice/js/config.js,apps/system/js/uuid.js,apps/bluetooth/js/settings.js,tests/atoms/remote_date.js,tests/atoms/screenshot.js,tests/performance/getappname_atom.js,tests/performance/exec-sync.js apps/system/js/notifications.js shared/js/lockscreen_slide.js 

jshint check:
apps/system/js/notifications.js: jshint errors are ignored. Please consider fixing lint errors and removing it from build/jshint/xfail.list.
shared/js/lockscreen_slide.js: jshint errors are ignored. Please consider fixing lint errors and removing it from build/jshint/xfail.list.
Please read https://github.com/mozilla-b2g/gaia/tree/master/build/jshint/README.md for more information.

apps/system/lockscreen/js/lockscreen.js: line 650, col 46, Missing semicolon. (ERROR)
apps/system/lockscreen/js/lockscreen_inputpad.js: line 137, col 96, Line is too long. (ERROR)
apps/system/lockscreen/js/lockscreen_inputpad.js: line 138, col 84, Line is too long. (ERROR)
apps/system/lockscreen/js/lockscreen_inputpad.js: line 146, col 14, Don'\''t make functions within a loop. (ERROR)
apps/system/lockscreen/js/lockscreen_inputpad.js: line 148, col 84, Line is too long. (ERROR)
apps/system/lockscreen/js/lockscreen_inputpad.js: line 171, col 82, Line is too long. (ERROR)
apps/system/lockscreen/js/lockscreen_inputpad.js: line 174, col 82, Line is too long. (ERROR)
apps/system/lockscreen/js/lockscreen_inputpad.js: line 181, col 7, Bad line breaking before '\''||'\''. (ERROR)
apps/system/lockscreen/js/lockscreen_inputpad.js: line 182, col 7, Bad line breaking before '\''||'\''. (ERROR)
apps/system/lockscreen/js/lockscreen_inputpad.js: line 183, col 8, Bad line breaking before '\''&&'\''. (ERROR)
apps/system/lockscreen/js/lockscreen_inputpad.js: line 189, col 4, Missing semicolon. (ERROR)
apps/system/lockscreen/js/lockscreen_notifications.js: line 302, col 107, Line is too long. (ERROR)
apps/system/lockscreen/js/lockscreen_state_manager.js: line 455, col 85, Line is too long. (ERROR)
apps/system/lockscreen/js/lockscreen_state_manager.js: line 624, col 4, Missing semicolon. (ERROR)

14 errors (81 xfailed)

TatianaMigacheva commented:
Harman-Jenkins: pr_review_failed

**Lint errors:**
gjslint check:
Running gjslint...
----- FILE  :  /export/jenkins/jobs/Gaia_PR_checker/workspace/checkstyle/gaia/apps/system/js/notifications.js -----
Line 437, E:0001: Extra space before ";"
Line 466, E:0002: Missing space after ","
Line 806, E:0002: Missing space before "+"
Line 820, E:0001: Extra space after "?"
----- FILE  :  /export/jenkins/jobs/Gaia_PR_checker/workspace/checkstyle/gaia/shared/js/lockscreen_slide.js -----
Line 671, E:0236: Error parsing jsdoc type at token "(" (column: 21): Invalid identifier for (): "Function"
Line 18, E:0002: Missing space before "|"
Line 18, E:0002: Missing space before "opts"
Line 212, E:0002: Missing space before "|"
Line 212, E:0002: Missing space before "opts"
Line 243, E:0002: Missing space before "|"
Line 243, E:0002: Missing space before "options"
Line 267, E:0002: Missing space before "|"
Line 267, E:0002: Missing space before "evt"
Line 474, E:0002: Missing space before "|"
Line 474, E:0002: Missing space before "tx"
Line 602, E:0002: Missing space before "|"
Line 602, E:0002: Missing space before "pageX"
Line 603, E:0002: Missing space before "|"
Line 603, E:0002: Missing space before "pageY"
Line 636, E:0002: Missing space before "|"
Line 636, E:0002: Missing space before "tx"
Line 670, E:0002: Missing space before "|"
Line 670, E:0002: Missing space before "tx"
Line 671, E:0002: Missing space before "|"
Line 671, E:0002: Missing space before "cb"
Line 737, E:0002: Missing space before "|"
Line 737, E:0002: Missing space before "tx"
Line 888, E:0002: Missing space before "|"
Line 888, E:0002: Missing space before "tx"
Line 1059, E:0002: Missing space before "|"
Line 1059, E:0002: Missing space before "px"
Line 1094, E:0002: Missing space before "|"
Line 1094, E:0002: Missing space before "px"
Line 1120, E:0002: Missing space before "|"
Line 1120, E:0002: Missing space before "x"
Line 1121, E:0002: Missing space before "|"
Line 1121, E:0002: Missing space before "y"
Found 37 errors, including 1 new error, in 2 files (0 files OK).

Some of the errors reported by GJsLint may be auto-fixable using the script
fixjsstyle. Please double check any changes it makes and report any bugs. The
script can be run by executing:

fixjsstyle --disable 210,217,220,225 --custom_jsdoc_tags=prop,borrows,memberof,augments,exports,global,event,example,mixes,mixin,fires,inner,todo,access,namespace,listens,module,memberOf,property,requires,alias,returns -e apps/pdfjs/content,apps/pdfjs/test,apps/email/build,apps/email/built,apps/email/js/ext,apps/calendar/build,apps/calendar/js/ext,apps/camera/js/vendor,apps/camera/test,apps/privacy-panel/js/vendor,apps/system/camera,apps/keyboard/js/imes/jszhuyin/lib,apps/keyboard/js/imes/handwriting/hwr,apps/wappush/js/ext,apps/keyboard/js/layouts,apps/music/bower_components,apps/camera/bower_components -x build/r.js,build/csslinter.js,build/jsmin.js,apps/communications/contacts/oauth2/js/parameters.js,apps/calendar/js/presets.js,apps/email/js/alameda.js,apps/email/js/tmpl_builder.js,apps/clock/js/alameda.js,apps/clock/js/text_builder.js,shared/js/utilities.js,shared/js/uuid.js,apps/keyboard/js/imes/jspinyin/libpinyin.js,shared/js/setImmediate.js,shared/test/unit/test-agent-server.js,shared/elements/gaia-header/dist/gaia-header.js,shared/elements/gaia-component/gaia-component.js,shared/elements/gaia-icons/gaia-icons.js,shared/pages/import/js/parameters.js,apps/sharedtest/test/unit/setImmediate_test.js,apps/keyboard/js/imes/jsavrophonetic/jsavrophonetic.js,apps/camera/js/vendor/alameda.js,apps/camera/js/config.js,apps/gallery/js/config.js,apps/gallery/js/frame_scripts.js,apps/gallery/js/metadata_scripts.js,apps/music/js/metadata_scripts.js,apps/keyboard/js/imes/jszhuyin/tools/cook.js,apps/keyboard/build/keyboard-config.js,apps/settings/js/vendor/alameda.js,apps/findmydevice/js/lib/*,apps/findmydevice/js/config.js,apps/system/js/uuid.js,apps/bluetooth/js/settings.js,tests/atoms/remote_date.js,tests/atoms/screenshot.js,tests/performance/getappname_atom.js,tests/performance/exec-sync.js apps/system/js/notifications.js shared/js/lockscreen_slide.js 

jshint check:
apps/system/js/notifications.js: jshint errors are ignored. Please consider fixing lint errors and removing it from build/jshint/xfail.list.
shared/js/lockscreen_slide.js: jshint errors are ignored. Please consider fixing lint errors and removing it from build/jshint/xfail.list.
Please read https://github.com/mozilla-b2g/gaia/tree/master/build/jshint/README.md for more information.

apps/system/lockscreen/js/lockscreen_inputpad.js: line 137, col 96, Line is too long. (ERROR)
apps/system/lockscreen/js/lockscreen_inputpad.js: line 138, col 84, Line is too long. (ERROR)
apps/system/lockscreen/js/lockscreen_inputpad.js: line 142, col 84, Line is too long. (ERROR)
apps/system/lockscreen/js/lockscreen_inputpad.js: line 174, col 82, Line is too long. (ERROR)
apps/system/lockscreen/js/lockscreen_inputpad.js: line 177, col 82, Line is too long. (ERROR)
apps/system/lockscreen/js/lockscreen_inputpad.js: line 190, col 4, Missing semicolon. (ERROR)

6 errors (68 xfailed)
There were errors while linting the files, please see above.
Please read https://github.com/mozilla-b2g/gaia/tree/master/build/jshint/README.md for more information.

harman-ankitgupta committed:
Bug 1210305 - Lockscreen: Implemented the lockscreen as per the UX.

Lock Screen UI implementation as per
http://r1l8an.axshare.com/#p=2_0_lock_screen

Modified Javascript files :-
- notifications.js
- lockscreen_notifications.js
As per the UX, similar kind of notifications should be clubbed together.
Tested by installing an external app and generated local notifications.

- lockscreen_inputpad.js
Password entered by the user on the lockscreen should be visible for the 1
second before changing it to dot.

- lockscreen_state_manager.js
If user clicks on the lockscreen notification and clicks on open after
unlocking the device , the corresponding app should be launched.

- Added visual design changes.
- Fixed the issues related to David's comment.
- Fixed the issue in passcode screen
- Fixed the issues related to Egor's comment.
- Changed the test cases as per the implementation
- As per the discussion with Egor, in case of lockscreen is on and
  passcode is off from settings, device will be unlocked from any key.
- Fixed the issue in lockscreen in which passcode screen is coming on
  the homescreen.
- Fixed the lint check errors.Errors such as variable used out of scope, variable is already defined are
not fixed as these are not real errors.

TatianaMigacheva commented:
Harman-Jenkins: pr_review_passed

TatianaMigacheva: Verified that Lint errors below exist in Mozilla Platform, so we will skip them.
**Lint errors:**

gjslint check:
Running gjslint...
----- FILE  :  /export/jenkins/jobs/Gaia_PR_checker/workspace/checkstyle/gaia/shared/js/lockscreen_slide.js -----
Line 671, E:0236: Error parsing jsdoc type at token "(" (column: 21): Invalid identifier for (): "Function"
Line 18, E:0002: Missing space before "|"
Line 18, E:0002: Missing space before "opts"
Line 212, E:0002: Missing space before "|"
Line 212, E:0002: Missing space before "opts"
Line 243, E:0002: Missing space before "|"
Line 243, E:0002: Missing space before "options"
Line 267, E:0002: Missing space before "|"
Line 267, E:0002: Missing space before "evt"
Line 474, E:0002: Missing space before "|"
Line 474, E:0002: Missing space before "tx"
Line 602, E:0002: Missing space before "|"
Line 602, E:0002: Missing space before "pageX"
Line 603, E:0002: Missing space before "|"
Line 603, E:0002: Missing space before "pageY"
Line 636, E:0002: Missing space before "|"
Line 636, E:0002: Missing space before "tx"
Line 670, E:0002: Missing space before "|"
Line 670, E:0002: Missing space before "tx"
Line 671, E:0002: Missing space before "|"
Line 671, E:0002: Missing space before "cb"
Line 737, E:0002: Missing space before "|"
Line 737, E:0002: Missing space before "tx"
Line 888, E:0002: Missing space before "|"
Line 888, E:0002: Missing space before "tx"
Line 1059, E:0002: Missing space before "|"
Line 1059, E:0002: Missing space before "px"
Line 1094, E:0002: Missing space before "|"
Line 1094, E:0002: Missing space before "px"
Line 1120, E:0002: Missing space before "|"
Line 1120, E:0002: Missing space before "x"
Line 1121, E:0002: Missing space before "|"
Line 1121, E:0002: Missing space before "y"
Found 33 errors, including 1 new error, in 1 file (1 file OK).

jshint check:

0 errors (68 xfailed)

**Css error:**
There were errors while linting the files, please see above.
Parsing errors in /export/build/ruintaka/2.2r_emulator_kk/B2G/gaia/apps/system/lockscreen/style/lockscreen.css
        Unknown property 'margin-inline-start'.  Declaration dropped. line: 210 column: 21 source: "  margin-inline-start: -0.4rem;"
        Unknown property 'margin-inline-end'.  Declaration dropped. line: 211 column: 19 source: "  margin-inline-end: 0;"
apps/system/lockscreen/style/lockscreen.css has new errors/warnings.

TatianaMigacheva commented:
TatianaMigacheva: gjslint errors and css errors exist in Mozilla platform so we will skip it in internal review.

harman-ankitgupta commented:
@TatianaMigacheva 

Errors are in existing code only, I have not changed anything related to these issues

harman-ankitgupta commented:
@egorlevichev 
Please review the changes.

egorlevichev commented:
@harman-red-square/harman-reviewers can anyone review and apply 'review' label for this PR?
